### PR TITLE
samjo's recipe-muncher

### PR DESCRIPTION
**Pull request by sjlee3157**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, and how did you try querying the API? | I used curl and Postman to help me explore the API. |
What is an API Wrapper? Why is it in `lib`? How would your project change if you needed to interact with more than one API (aka more than just the Edamam API)? | The API Wrapper limits the interface of the Edamam API and my app in order to isolate dependency. I think it's in lib because it's not part of the Rails MVC architecture. My app does interact with more than one API if you count OAuth strategies and the Google People API. Omniauth is doing most of the "wrapper" work for me, but if it wasn't, I would be creating another wrapper class and figuring out how to parse the JSON/XML responses. |
Describe your API wrapper, the methods you created in it, and why you decided to create those methods/how they were helpful | I created these methods: self.list_recipes, self.get_recipe, self.list_tags, and the self.build_recipe factory method. The methods were necessitated by the structure of the Edamam API, which just happens to accept GET requests formatted in specific ways that I addressed in my methods. |
What was an edge case or failure case test you wrote for your API Wrapper? What was a nominal case? | Edge: trying to "get" a result from Edamam API with a bogus recipe id. Nominal: trying to get a result with a valid recipe id. |
How does VCR aid in testing an API? | VCR records and plays back API calls which is useful because API calls are often limited, expensive, and unreliable! |
What is the Heroku URL of your deployed application? | https://hello-munchy.herokuapp.com |

Trello: https://trello.com/b/HL0gtltd/sj-api-muncher
See Trello for what remains to be done. I'm not finished with testing, but a lot of what's left is testing that I've practiced with MR-R and bEtsy. While TDD is a good practice, for this project I decided to prioritize learning new skills, like OAuth Google and filtering search results.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+Rails.application.routes.draw do
+  root 'recipe_searches#home'
+
+  get "/auth/:provider/callback", to: "sessions#create", as: 'auth_callback'
+  delete "/logout", to: "sessions#destroy", as: "logout"
+
+  resources :recipe_searches, only: [:index, :show]
```
Comment (dHelmgren) [inline config/routes.rb:9]: Nice work setting these with `resources`

File: lib/recipe.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class Recipe
+  attr_reader :label, :recipe_uri, :source, :source_uri, :ingredient_lines,
+              :image_uri, :recipe_yield, :total_time, :health_labels
+
+  REQUIRED_ARGS = [:label, :recipe_uri, :source, :source_uri, :ingredient_lines]
```
Comment (dHelmgren) [inline lib/recipe.rb:5]: I would say you're over-validating here. In general, you should only validate fields that will cause your app to break if they're not there. In this case, that's `label`, `uri` and possibly `image`. Anything beyond that is an unnecessary dependency on the data coming from the API and is an extra thing for you to test.

File: test/cassettes/get_recipe_file.yml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      X-Served-By:
+      - ip-10-140-213-39/10.140.213.39
+    body:
+      encoding: UTF-8
+      string: "[]"
+    http_version: 
+  recorded_at: Wed, 31 Oct 2018 22:22:23 GMT
+- request:
+    method: get
+    uri: https://api.edamam.com/search?app_id=<EDAMAM_APP_ID>&app_key=<EDAMAM_APP_KEY>&r=http://www.edamam.com/ontologies/edamam.owl%23bogus!
```
Comment (dHelmgren) [inline test/cassettes/get_recipe_file.yml:90]: yay!

File: test/controllers/favorites_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
+require "test_helper"
+
+describe FavoritesController do
+  it "should get index" do
+    get favorites_path
```
Comment (dHelmgren) [inline test/controllers/favorites_controller_test.rb:5]: There are some interesting cases you're not covering here:

- What happens if no search term is provided?
- What if the search returns no results?

It might also be worthwhile to add some tests around the paging parameters:

- What happens if they're absent?
- Do you get different results when they change?
- What if you send a bogus value, like a negative page number?

Just FYI, Your pagination breaks if I give it a negative page.

**Comment by dHelmgren** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene | yes
Comprehension questions | good
**General** |
Rails fundamentals (RESTful routing, use of named paths) | Yes!
Semantic HTML | That Index page has a lotta divs.
Errors are reported to the user | Mostly see comment on Index testing.
API Wrapper to handle the API requests | yes
Controller testing | Pretty good see comments
Lib testing | good!
Search Functionality | yes
List Functionality | yes
Show individual item functionality | yes
**Styling** |
List view shows 10 items at a time and/or has pagination | yes
Attractive and usable UI | so so. I think the flippy recipe cards are great, I had a hard time spotting your search bar.
**API Features** |
The App attributes Edamam | You need to attribute it on every page. This is really important! When you're designing code for money, using someone else's work without attribution is a good way to end up in legal trouble.
The VCR cassettes do not contain the API key | yes!
**External Resources** |
Link to deployed app on Heroku | Yes
**Overall** | This is a good showing! You hit the learning goals very well, just make sure that your Attributions are complete in the future.

File: lib/recipe.rb (hunk 2 above)
Comment (sjlee3157) [reply lib/recipe.rb]: for RUBY!
| Row | Function | Hash | Set | Array |
| --- | -------- | ------ | ---- | ----- |
| 1. | insert | O(1) avg / O(n) worst | (merge) O(1) avg/ O(n) worst | append: O(1) avg / O(n) worst if dynamic, prepend: O(1) dynamic, O(n) if static |
| 3. | size (if OO implementation tracking size) | O(1) | O(1) | O(1) |
| 4. | lookup value by key/index | O(1) / O(n) worst | O(1) | O(1) |
| 5. | .include?() OR lookup key/index by value | O(n) | O(1) | O(n) |
